File: rust/src/0053_maximum_subarray.rs
pub struct Solution;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        nums.iter()
            .scan(i32::MIN, |state, &x| {
                *state = if *state <= 0 { x } else { *state + x };
                Some(*state)
            })
            .inspect(|x| println!("{}", x))
            .max()
            .unwrap()
    }
}

fn main() {
    println!("{}", Solution::max_sub_array(vec![-2,1,-3,4,-1,2,1,-5,4]))
}

File: 0053_MaximumSubarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int best_sum = nums[0];
        int current_sum = 0;

        for (int x : nums) {
            current_sum = Math.max(x, current_sum + x);
            best_sum = Math.max(best_sum, current_sum);
        }

        return best_sum;
    }

    public static void main(String[] args) {
        int[] nums = {-2,1,-3,4,-1,2,1,-5,4};
        Solution sol = new Solution();
        System.out.println(sol.maxSubArray(nums));
    }
}
